### ISSUE TITLE
Jitting Callback Function

### ISSUE THREAD

**Issue by sgstepaniants** [patrick-kidger/equinox#991]
I have a question concerning speeding up my callback function which plots several solutions to a neural ODE. Because evaluating the neural ODE model takes a bit of time (and I make several plots), the callback function ends up taking more time to plot 3 neural ODE predictions than it takes to run through one batch of 30 trajectories in a training step. I think this is because my callback function or model is not jitted. Sticking with equinox and diffrax best practices, I am currently only jitting the functions "make_step" and "grad_loss". So currently my callback function which I call after every batch training step is unjitted and looks like this:

`def callback(steps, key, selected_inds=jnp.array([]), num_random=0):
    ymin = min(y_full)
    ymax = max(y_full)
    
    random_inds = jnp.arange(y_train.shape[0])
    random_inds = jnp.setdiff1d(random_inds, selected_inds)
    random_inds = jr.choice(key, random_inds, shape=(num_random,), replace=False)
    inds = jnp.concatenate([selected_inds, random_inds])
    
    fig, axes = plt.subplots(1, len(inds), figsize=(20, 4))
    for idx, ax in enumerate(axes):
        i = int(inds[idx])
        tpred, ypred = model(steps, y_train[i, :, None])
        "some plotting code ..."
    
    plt.tight_layout()
    plt.show()`

**Comment by patrick-kidger** [comment]
You should separate your numerical JAX bit and the general 'software' bit of plotting. Do all JAX operations inside a jit'd region, then pass their output to the rest of your program.

**Comment by sgstepaniants** [comment]
I see, so the only numerical operation in the callback is calling my equinox model "__call__" function in a for loop. Should I jit the "__call__" function that is defined in the equinox model class? I didn't see any examples doing this so I avoided taking this route.

**Comment by johannahaffner** [comment]
You could decorate `__call__` with `jit`, but more commonly you instead want to jit your loss function and `make_step` function that applies the optimiser updates. 

For plotting, you could create an outer loop that is just Python - looping over epochs - which then has an inner loop over batches. At the end of this outer loop, you can extract some value and create a plot. This setup allows you to compile the expensive stuff and still generate plots, without requiring the use of a callback.

**Comment by johannahaffner** [comment]
Yes, that sounds like more trouble than it's worth! You could probably do something like:

```python
@eqx.filter_jit 
def loss(model, data):
      ...

@eqx.filter_jit
def make_step(...)
      ...

for epoch in epochs:
      # Train across batches, get updated model
      current_loss = loss(updated_model, data)
      # Either plot something or just record the loss to plot something later
```

And so on. The plotting function would anyway not be something for which you can expect any kind of performance gain by placing it in a jitted region with a callback-workaround. I never do this - and I'm not even sure it could be done reliably: https://docs.jax.dev/en/latest/external-callbacks.html#flavors-of-callback

**Comment by ak24watch** [comment]
> Yes, that sounds like more trouble than it's worth! You could probably do something like:
> 
> @eqx.filter_jit 
> def loss(model, data):
>       ...
> 
> @eqx.filter_jit
> def make_step(...)
>       ...
> 
> for epoch in epochs:
>       # Train across batches, get updated model (this is where make_step gets called)
>       current_loss = loss(updated_model, data)
>       # Either plot something or just record the loss to plot something later
> And so on. The plotting function would anyway not be something for which you can expect any kind of performance gain by placing it in a jitted region with a callback-workaround. I never do this - and I'm not even sure it could be done reliably: https://docs.jax.dev/en/latest/external-callbacks.html#flavors-of-callback






> **i think jitting loss function is not necessary as  loss function call is inside make_step and make step is jitted.  it will be twice jitted if done so , even if want to jit loss function then it must be jax.jit(jax.grad) rather than jax.grad(jax.jit)**  
> **i think jitting outermost function is necessary for optimal performance**
